File: Graphs/Algorithms/Flow Algoritms/Dinic.py
from math import inf
from collections import deque

class Dinic:
    def __init__(self, n):
        self.n = n
        self.adj = [[] for _ in range(n)]  # each edge: [to, rev_index, cap]
        self.level = [0]*n
        self.it = [0]*n

    def add_edge(self, u, v, cap):
        rev_u = len(self.adj[v])
        self.adj[u].append([v, rev_u, cap])
        rev_v = len(self.adj[u]) - 1
        self.adj[v].append([u, rev_v, 0])
        return rev_v  # index of forward edge in adj[u]

    def bfs(self, s, t):
        self.level = [-1]*self.n
        dq = deque([s]); self.level[s]=0
        while dq:
            v = dq.popleft()
            for to, _, cap in self.adj[v]:
                if cap>0 and self.level[to]<0:
                    self.level[to]=self.level[v]+1
                    dq.append(to)
                    
        return self.level[t]>=0

    def dfs(self, v, t, pushed):
        if v==t or pushed==0: return pushed
        for i in range(self.it[v], len(self.adj[v])):
            self.it[v]=i
            to, rev, cap = self.adj[v][i]
            if cap==0 or self.level[to]!=self.level[v]+1: continue
            tr = self.dfs(to, t, min(pushed, cap))
            if tr==0: continue
            self.adj[v][i][2] -= tr
            self.adj[to][rev][2] += tr
            return tr
        return 0

    def max_flow(self, s, t):
        flow = 0
        while self.bfs(s,t):
            self.it = [0]*self.n
            while True:
                pushed = self.dfs(s,t, inf)
                if pushed==0: break
                flow += pushed
        return flow

n = 4
d = Dinic(n)
edges = []  # (u,v,cap, idx_in_adj_u)
def add_edge(u,v,cap):
    idx = d.add_edge(u,v,cap)
    edges.append((u,v,cap,idx))
add_edge(0,1,3)
add_edge(0,2,2)
add_edge(1,2,1)
add_edge(1,3,2)
add_edge(2,3,4)

s, t = 0, 3
flow = d.max_flow(s,t)
print(f"Max flow from {s} to {t} = {flow}\n")

print("Edge flows (u -> v):")
for u,v,cap,idx in edges:
    remaining = d.adj[u][idx][2]
    sent = cap - remaining
    print(f"{u} -> {v} : sent = {sent} (residual cap = {remaining})")

print("\nResidual capacity matrix:")
res = [[0]*n for _ in range(n)]
for u in range(n):
    for to, _, cap in d.adj[u]:
        res[u][to] += cap
for row in res:
    print(' '.join(str(x) for x in row))

File: Graphs/Algorithms/Flow Algoritms/Dinic.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;
typedef long double ld;
const ll INF = numeric_limits<ll>::max();

#define all(x) x.begin(), x.end()
#define rall(x) x.rbegin(), x.rend()
#define MOD 998244353
// #define MOD pow(10, 9)+7;
#define pi acos(-1)
#define gcd __gcd
#define lcm(a, b) ((a) * (b) / gcd(a, b))
#define precision(n) cout << fixed << setprecision(n)
#define debug(arr) cerr << #arr << " = "; for (auto &x : arr) cerr << x << " "; cerr << endl;
#define fastio ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);

struct Dinic {
    struct Edge { 
        ll to; size_t rev_index; ll capacity;            
    };

    ll n;
    vector<vector<Edge>> adj;
    vector<ll> level;
    vector<ll> it;

    Dinic(ll n=0): n(n), adj(n), level(n), it(n) {}

    ll add_edge(ll u, ll v, ll cap){
        Edge a{v, adj[v].size(), cap};
        Edge b{u, adj[u].size(), 0};
        adj[u].push_back(a);
        adj[v].push_back(b);
        return adj[u].size() - 1; // index of forward edge in adj[u]
    }

    bool bfs(ll s, ll t){
        fill(all(level), -1); 
        level[s] = 0;

        queue<ll> q; q.push(s);
        while(!q.empty()){
            int v = q.front(); q.pop();
            for(auto &edge : adj[v]){
                if(level[edge.to] < 0 && edge.capacity > 0){
                    level[edge.to] = level[v] + 1;
                    q.push(edge.to);
                }
            }
        }
        return level[t] >= 0;
    }

    ll dfs(ll v, ll t, ll pushed){
        if(v == t || pushed == 0) return pushed;

        for(ll i = it[v]; i < adj[v].size(); ++i){
            Edge &edge = adj[v][i];

            if(level[edge.to] != level[v] + 1 || edge.capacity == 0) continue;
            
            ll tr = dfs(edge.to, t, min(pushed, edge.capacity));
            if(tr == 0) continue;

            edge.capacity -= tr;
            adj[edge.to][edge.rev_index].capacity += tr;
            return tr;
        }
        return 0;
    }

    ll max_flow(ll s, ll t){
        ll flow = 0;
        while(bfs(s,t)){
            fill(it.begin(), it.end(), 0);
            while(ll pushed = dfs(s,t, INF)) flow += pushed;
        }
        return flow;
    }
};

int main(){
    fastio;
    // Example graph (same as before)
    // 0 -> 1 (cap 3)
    // 0 -> 2 (cap 2)
    // 1 -> 2 (cap 1)
    // 1 -> 3 (cap 2)
    // 2 -> 3 (cap 4)
    //
    // Max flow from 0 to 3 should be 5.

    ll n = 4;
    Dinic dinic(n);
    vector<tuple<ll,ll,ll,ll>> edges; // (u,v,cap, idx_in_adj_u)

    auto add_edge = [&](ll u, ll v, ll cap){
        int idx = dinic.add_edge(u, v, cap);
        edges.emplace_back(u, v, cap, idx);
    };

    add_edge(0, 1, 3);
    add_edge(0, 2, 2);
    add_edge(1, 2, 1);
    add_edge(1, 3, 2);
    add_edge(2, 3, 4);

    // keep a copy of original capacities to compute flow on each edge later
    vector<vector<ll>> capacity_init(n, vector<ll>(n, 0));
    for(auto &t : edges){
        int u,v; ll cap; int idx; tie(u,v,cap,idx) = t;
        capacity_init[u][v] = cap;
    }

    int s = 0, t = 3;
    ll flow = dinic.max_flow(s, t);
    cout << "Max flow from " << s << " to " << t << " = " << flow << '\n';

    // Print flow sent on each original edge and residual capacity
    cout << "Edge flows (u -> v):\n";
    for (auto &e : edges){
        int u, v; ll cap; int idx;
        tie(u, v, cap, idx) = e;
        ll remaining = dinic.adj[u][idx].capacity;         // remaining capacity on forward edge
        ll sent = cap - remaining;                    // flow sent
        cout << u << " -> " << v << " : sent = " << sent 
             << " (residual cap = " << remaining << ")\n";
    }

    // Print full residual capacity matrix
    cout << "Residual capacity matrix:\n";
    vector<vector<ll>> residual(n, vector<ll>(n, 0));
    for(int u = 0; u < n; ++u){
        for(auto &ed : dinic.adj[u]){
            residual[u][ed.to] += ed.capacity; // note: multiple edges could exist
        }
    }
    for(int i = 0; i < n; ++i){
        for(int j = 0; j < n; ++j){
            cout << residual[i][j] << (j+1==n?'\n':' ');
        }
    }

    return 0;
}